rust-runtime/smithy-eventstream/src/frame.rs

```diff
@@ -0,0 +1,590 @@
+        let headers: arbitrary::Result<Vec<Header>> = unstruct.arbitrary_iter()?.collect();
+        Ok(Message {
+            headers: headers?,
+            payload: Bytes::from(Vec::<u8>::arbitrary(unstruct)?),
+        })
+    }
+}
+
+fn checked<T: TryFrom<U>, U>(from: U, err: Error) -> Result<T, Error> {
+    T::try_from(from).map_err(|_| err)
+}
+
+fn max_header_len(total_len: u32) -> Result<u32, Error> {
+    total_len
+        .checked_sub(PRELUDE_LENGTH_BYTES + MESSAGE_CRC_LENGTH_BYTES)
+        .ok_or(Error::InvalidMessageLength)
+}
+
+fn payload_len(total_len: u32, header_len: u32) -> Result<u32, Error> {
+    total_len
+        .checked_sub(
+            header_len
+                .checked_add(PRELUDE_LENGTH_BYTES + MESSAGE_CRC_LENGTH_BYTES)
+                .ok_or(Error::InvalidHeadersLength)?,
+        )
+        .ok_or(Error::InvalidMessageLength)
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::error::Error;
+    use crate::frame::{Header, HeaderValue, Message};
+    use bytes::Bytes;
+    use smithy_types::Instant;
+
+    macro_rules! read_message_expect_err {
+        ($bytes:expr, $err:pat) => {
+            let result = Message::read_from(&mut Bytes::from_static($bytes));
+            assert!(
+                matches!(&result.as_ref(), &Err($err)),
+                "Expected {}, got {:?}",
+                stringify!(Err($err)),
+                result
+            );
+        };
+    }
+
+    #[test]
+    fn invalid_messages() {
+        read_message_expect_err!(
+            include_bytes!("../test_data/invalid_header_string_value_length"),
+            Error::InvalidHeaderValue
+        );
+        read_message_expect_err!(
+            include_bytes!("../test_data/invalid_header_string_length_cut_off"),
+            Error::InvalidHeaderValue
+        );
+        read_message_expect_err!(
+            include_bytes!("../test_data/invalid_header_value_type"),
+            Error::InvalidHeaderValueType(0x60)
+        );
+        read_message_expect_err!(
+            include_bytes!("../test_data/invalid_header_name_length"),
+            Error::InvalidHeaderNameLength
+        );
+        read_message_expect_err!(
+            include_bytes!("../test_data/invalid_headers_length"),
+            Error::InvalidHeadersLength
+        );
+        read_message_expect_err!(
+            include_bytes!("../test_data/invalid_prelude_checksum"),
+            Error::PreludeChecksumMismatch(0x8BB495FB, 0xDEADBEEF)
+        );
+        read_message_expect_err!(
+            include_bytes!("../test_data/invalid_message_checksum"),
+            Error::MessageChecksumMismatch(0x01a05860, 0xDEADBEEF)
+        );
+        read_message_expect_err!(
+            include_bytes!("../test_data/invalid_header_name_length_too_long"),
+            Error::InvalidHeaderValue
+        );
+    }
+
+    #[test]
+    fn message_not_fully_available_yet() {
+        let message = include_bytes!("../test_data/valid_with_all_headers_and_payload");
```

Comment:
how are callers supposed to deal with the partially mutated buffer that results from a call to `read_from` with a partial buffer? It seems like it actually reads through the prelude right? So you couldn't re-use the same bytes later.

would be good to formalize this behavior and document how it should be used in the context of something that is `impl Stream<Bytes>`

Reply:
I imagine the caller must have to do something to preserve the initial position in the event it returns `None`. There's no peeking with `Buf`, so there is no other way to do that. I should double check this though instead of assuming.